File: sdk/feedback/kotlin/src/test/kotlin/io/yaver/feedback/ReloadActionsTest.kt
package io.yaver.feedback

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Unit tests for the PURE reload-decision seam.
 *
 * These need no emulator and no Android framework — [ReloadActions] touches
 * neither, which is the whole reason it is a separate file from
 * [YaverReloadControl].
 *
 * ── NOT YET EXECUTED ─────────────────────────────────────────────────────────
 *
 * `sdk/feedback/kotlin/` has NO build harness — no build.gradle, no wrapper,
 * and there is no kotlinc on the machine this was written on, so this file has
 * never been run. It is committed anyway so that the guard exists the moment a
 * harness does, but it must not be counted as passing until someone has
 * watched it fail. The identical logic IS executed today in the web, RN,
 * Flutter and Swift ports, whose equivalents of
 * [releaseBuildGetsNoReloadActionsAtAll] have each been verified by breaking
 * the guard and watching the suite go red.
 */
class ReloadActionsTest {

    private fun running(framework: String) =
        DevServerSnapshot(running = true, framework = framework)

    // ─── THE GUARD ───────────────────────────────────────────────────────────
    //
    // Prove it by breaking it: change `if (!isDevBuild) return emptyList()` in
    // ReloadActions.build to `if (isDevBuild)` and this single test fails
    // while every other test in this file still passes.
    @Test
    fun releaseBuildGetsNoReloadActionsAtAll() {
        val actions = ReloadActions.build(running("expo"), isDevBuild = false, connected = true)

        assertTrue(actions.isEmpty())
    }

    @Test
    fun debugBuildGetsHotAndFull() {
        val actions = ReloadActions.build(running("expo"), isDevBuild = true, connected = true)

        assertEquals(listOf(ReloadActionId.HOT, ReloadActionId.FULL), actions.map { it.id })
        assertTrue(actions.all { it.enabled })
    }

    @Test
    fun frameworkFamilyMapsTheAgentNames() {
        assertEquals(ReloadFrameworkFamily.FLUTTER, ReloadActions.frameworkFamily("flutter"))
        assertEquals(ReloadFrameworkFamily.REACT_NATIVE, ReloadActions.frameworkFamily("expo"))
        assertEquals(ReloadFrameworkFamily.REACT_NATIVE, ReloadActions.frameworkFamily("react-native"))
        assertEquals(ReloadFrameworkFamily.WEB, ReloadActions.frameworkFamily("vite"))
        assertEquals(ReloadFrameworkFamily.WEB, ReloadActions.frameworkFamily("nextjs"))
        assertEquals(ReloadFrameworkFamily.UNKNOWN, ReloadActions.frameworkFamily(""))
        assertEquals(ReloadFrameworkFamily.UNKNOWN, ReloadActions.frameworkFamily("godot"))
    }

    @Test
    fun flutterSecondActionIsAHotRestartNotAFullReload() {
        val actions = ReloadActions.build(running("flutter"), isDevBuild = true, connected = true)

        assertEquals("Hot Reload", actions[0].label)
        assertEquals("Hot Restart", actions[1].label)
        assertTrue(actions[1].hint.contains("(R)"))
    }

    @Test
    fun everyOtherFrameworkCallsItAFullReload() {
        for (framework in listOf("expo", "vite", "nextjs")) {
            val actions =
                ReloadActions.build(running(framework), isDevBuild = true, connected = true)
            assertEquals(framework, "Full Reload", actions[1].label)
        }
    }

    @Test
    fun payloadIsFastThenFullBothOnDevReload() {
        val actions = ReloadActions.build(running("expo"), isDevBuild = true, connected = true)

        assertEquals("/dev/reload", actions[0].path)
        assertEquals("""{"mode":"fast"}""", actions[0].bodyJson)
        assertEquals("/dev/reload", actions[1].path)
        assertEquals("""{"mode":"full"}""", actions[1].bodyJson)
    }

    @Test
    fun neverOffersTheHermesBundlePathAKotlinAppCannotLoad() {
        val actions = ReloadActions.build(running("expo"), isDevBuild = true, connected = true)

        assertFalse(actions.any { it.path == ReloadActions.RELOAD_APP_PATH })
    }

    @Test
    fun noDevServerNamesTheMachineAndTheCommandThatStartsIt() {
        val actions = ReloadActions.build(
            DevServerSnapshot(running = false),
            isDevBuild = true,
            connected = true,
            machineLabel = "primary",
        )

        for (action in actions) {
            assertFalse(action.enabled)
            assertTrue(action.disabledReason!!.contains("primary"))
            assertTrue(action.disabledReason!!.contains("yaver dev start"))
        }
    }

    @Test
    fun buildingSaysStillBuildingNotNoDevServer() {
        val actions = ReloadActions.build(
            DevServerSnapshot(running = true, building = true, framework = "expo"),
            isDevBuild = true,
            connected = true,
        )

        assertTrue(actions[0].disabledReason!!.contains("still building"))
    }

    @Test
    fun disconnectedSaysNotConnectedNotNoDevServer() {
        val actions = ReloadActions.build(running("expo"), isDevBuild = true, connected = false)

        assertTrue(actions[0].disabledReason!!.contains("Not connected"))
    }

    @Test
    fun snapshotFromJsonDegradesToNotRunningNeverOptimism() {
        val live = DevServerSnapshot.fromJson(
            JSONObject("""{"running":true,"framework":"expo"}""")
        )
        assertTrue(live.running)
        assertEquals("expo", live.framework)

        val empty = DevServerSnapshot.fromJson(JSONObject("{}"))
        assertFalse(empty.running)
        assertFalse(empty.building)
        assertNull(empty.framework)
    }

    @Test
    fun describeFailureNamesACauseNeverJustFailed() {
        assertTrue(
            ReloadActions.describeFailure(503, "dev server not available")
                .contains("No dev server is running")
        )
        assertTrue(
            ReloadActions.describeFailure(500, "vite does not support hot reload", running("vite"))
                .contains("vite")
        )
        assertTrue(
            ReloadActions.describeFailure(
                502,
                "Get \"http://127.0.0.1:8081/reload\": dial tcp 127.0.0.1:8081: connect: connection refused"
            ).contains("not listening")
        )
        assertTrue(ReloadActions.describeFailure(401, "").contains("sign in again"))
        assertTrue(ReloadActions.describeFailure(403, "").contains("sign in again"))
        assertTrue(ReloadActions.describeFailure(404, "not found").contains("yaver-cli@latest"))
        assertTrue(ReloadActions.describeFailure(500, "boom").contains("yaver logs"))
        assertTrue(ReloadActions.describeFailure(0, "").contains("yaver serve"))
    }
}

File: sdk/feedback/browser-extension/test/reloadActions.test.js
/* eslint-disable no-undef */
const test = require('node:test');
const assert = require('node:assert');

const {
  RELOAD_APP_PATH,
  RELOAD_PATH,
  describeReloadFailure,
  isDevAgentUrl,
  reloadActions,
  reloadFrameworkFamily,
  reloadRequest,
} = require('../reloadActions.js');

const DEV = { isDevBuild: true, connected: true };
const running = (framework) => ({ running: true, framework });

// ─── THE GUARD ───────────────────────────────────────────────────────────────
//
// Prove it by breaking it: change `if (!options.isDevBuild) return [];` in
// reloadActions.js to `if (options.isDevBuild) return [];` and this single
// test fails while every other test in this file still passes.
test('a non-dev context gets NOTHING, even with a healthy dev server', () => {
  const actions = reloadActions(running('vite'), { isDevBuild: false, connected: true });
  assert.deepStrictEqual(actions, []);
});

test('a dev context gets hot + full', () => {
  const actions = reloadActions(running('vite'), DEV);
  assert.deepStrictEqual(actions.map((a) => a.id), ['hot', 'full']);
  assert.ok(actions.every((a) => a.enabled));
});

// ─── isDevAgentUrl is the extension's version of "is this a dev build" ───────
//
// manifest.json grants host_permissions for localhost + 127.0.0.1 ONLY, so a
// public agent URL cannot work — offering reload buttons for one would be
// offering an action that cannot succeed.
test('isDevAgentUrl accepts only hosts the manifest actually permits', () => {
  assert.strictEqual(isDevAgentUrl('http://localhost:18080'), true);
  assert.strictEqual(isDevAgentUrl('http://127.0.0.1:18080'), true);
  assert.strictEqual(isDevAgentUrl('https://localhost:18443'), true);
  assert.strictEqual(isDevAgentUrl('http://box.localhost:18080'), true);

  assert.strictEqual(isDevAgentUrl('https://public.yaver.io/d/abc'), false);
  assert.strictEqual(isDevAgentUrl('http://192.168.1.42:18080'), false);
  assert.strictEqual(isDevAgentUrl(''), false);
  assert.strictEqual(isDevAgentUrl('not a url'), false);
  assert.strictEqual(isDevAgentUrl(undefined), false);
});

test('reloadFrameworkFamily maps the agent framework names', () => {
  assert.strictEqual(reloadFrameworkFamily('flutter'), 'flutter');
  assert.strictEqual(reloadFrameworkFamily('expo'), 'react-native');
  assert.strictEqual(reloadFrameworkFamily('react-native'), 'react-native');
  assert.strictEqual(reloadFrameworkFamily('vite'), 'web');
  assert.strictEqual(reloadFrameworkFamily('nextjs'), 'web');
  assert.strictEqual(reloadFrameworkFamily(''), 'unknown');
  assert.strictEqual(reloadFrameworkFamily('godot'), 'unknown');
});

test('Flutter second action is a Hot Restart, everyone else a Full Reload', () => {
  const flutter = reloadActions(running('flutter'), DEV);
  assert.strictEqual(flutter[0].label, 'Hot Reload');
  assert.strictEqual(flutter[1].label, 'Hot Restart');
  assert.ok(flutter[1].hint.includes('(R)'));

  for (const framework of ['expo', 'vite', 'nextjs']) {
    assert.strictEqual(reloadActions(running(framework), DEV)[1].label, 'Full Reload', framework);
  }
});

test('URL / payload construction', () => {
  const actions = reloadActions(running('vite'), DEV);
  assert.deepStrictEqual(reloadRequest(actions[0]), {
    method: 'POST',
    path: RELOAD_PATH,
    body: { mode: 'fast' },
  });
  assert.deepStrictEqual(reloadRequest(actions[1]), {
    method: 'POST',
    path: RELOAD_PATH,
    body: { mode: 'full' },
  });
});

test('never offers the Hermes bundle path from a browser extension', () => {
  const actions = reloadActions(running('react-native'), DEV);
  assert.ok(!actions.some((a) => a.path === RELOAD_APP_PATH));
});

test('a blocked action NAMES the blocker', () => {
  const noServer = reloadActions({ running: false }, { ...DEV, machineLabel: 'localhost:18080' });
  for (const action of noServer) {
    assert.strictEqual(action.enabled, false);
    assert.ok(action.disabledReason.includes('localhost:18080'));
    assert.ok(action.disabledReason.includes('yaver dev start'));
  }

  const building = reloadActions({ running: true, building: true, framework: 'vite' }, DEV);
  assert.ok(building[0].disabledReason.includes('still building'));

  const offline = reloadActions(running('vite'), { isDevBuild: true, connected: false });
  assert.ok(offline[0].disabledReason.includes('Not connected'));
});

test('describeReloadFailure names a cause, never just "failed"', () => {
  assert.ok(describeReloadFailure(503, 'dev server not available').includes('No dev server is running'));
  assert.ok(
    describeReloadFailure(500, 'vite does not support hot reload', running('vite')).includes('vite'),
  );
  assert.ok(
    describeReloadFailure(
      502,
      'Get "http://127.0.0.1:5173/reload": dial tcp 127.0.0.1:5173: connect: connection refused',
    ).includes('not listening'),
  );
  assert.ok(describeReloadFailure(401, '').includes('auth token'));
  assert.ok(describeReloadFailure(403, '').includes('auth token'));
  assert.ok(describeReloadFailure(404, 'not found').includes('yaver-cli@latest'));
  assert.ok(describeReloadFailure(500, 'boom').includes('yaver logs'));
  assert.ok(describeReloadFailure(0, '').includes('yaver serve'));
});
